the-keelson-protocol.md

```diff
@@ -0,0 +1,87 @@
+  * `source_id` being a unique id for the source producing the information described by `subject`. `source_id` may contain any number of addititional levels (i.e. forward slashes `/`)
+
+### 2.2 Message format specification
+
+Each message published to zenoh must be a protobuf-encoded keelson `Envelope`. An `Envelope` contains exactly one (1) `payload`, we say that a `payload` is **enclosed** within an `Envelope` by the publisher and can later be **uncovered** from that `Envelope` by the subscriber. 
+
+![sketch](subject_payload_schema.drawio.svg)
+
+Keelson support a set of well-known `payload`s, defined by the protobuf schemas available in [messages](./messages/payloads/). Each well-known `payload` is associated with an informative `subject`, the mapping between `subject`s and `payload`s is maintained in a [look-up table in YAML format](./messages/subjects.yaml).
+
+The main design principles behind this scheme are:
+* Well-known payloads are defined by a schema that describes how to interpret the **data**.
+* Each (well-known) payload is associated with a subject that describes how to interpret the **information**.
+* Each subject is part of the key when publishing data to zenoh, refer to the section about [keys](#21-specific-key-space-design), this helps the sender and receiver to put the information into a **context**.
+
+#### 2.2.1 Naming convention for `subject`s
+
+There are three distinct kind of payloads that has to be covered by a naming convention for `subject`s:
+* **arbitrary bytes**, where we dont know the schema or dont want to express the schema as a protobuf type, these all fall under the special subject `raw` using the payload type [`TimestampedBytes`](./messages/payloads/TimestampedBytes.proto)
+* **primitive payloads**, which have a specific meaning but where the protobuf type is generic, i.e [`TimestampedFloat`](./messages/payloads/TimestampedFloat.proto) or similar. In this case the subject needs to be very informative with regards to that value and we employ the following convention: `<entity>_<property>_<unit>` where `entity`, `property` and `unit` are constrained to alphanumeric characters. For example `rudder_angle_deg`.
+* **complex payloads**, which have a specific protobuf type that is not shared with any other subject. In this case, the subject name should be the snake_case version of the protobuf message name, for example `RawImage` -> `raw_image`.
+
+In general, [`subjects.yaml`](./messages/subjects.yaml) contains the current well-known subjects and can be regarded as the style-guide to follow.
+
+## 3. Request-Reply messaging (Remote Procedure Calls (RPC))
+
+### 3.1 Specific key-space design
+
+For the req/rep messaging pattern, the lower level hierarchy in the key space consists of the following levels:
+
+  `.../rpc/{responder_id}/{procedure}`
```

Comment:
The specifications of the protobuf services are located in the `interfaces` folder. As with "data" in section 2.1 above, I think it is suitable to match the name of the folder and descriptive word in the URL. If you look inside the `connectors` folder, there are other READMEs that contain the word "interface"; thus, the contents of the `connectors` folder are interfaces, and, in my mind, the contents of the `interfaces` folder are remote procedure calls (rpc) or "services". I suggest that the `connectors` folder is renamed to `interfaces` and the `interfaces` folder is renamed to `rpc`. Then we would get that nice match between the URL and the naming of the folders.

Reply:
Ok. Sold.